portal.go

```diff
@@ -2047,22 +2063,65 @@ func (portal *Portal) HandleiMessage(msg *imessage.Message) id.EventID {
 		portal.bridge.IM.SendMessageBridgeResult(msg.ChatGUID, msg.GUID, eventID, overrideSuccess || hasMXID)
 	}()
 
+	// Look up the message in the database
+	dbMessage = portal.bridge.DB.Message.GetLastByGUID(portal.GUID, msg.GUID)
+
 	if portal.IsPrivateChat() && msg.ChatGUID != portal.LastSeenHandle {
 		portal.log.Debugfln("Updating last seen handle from %s to %s", portal.LastSeenHandle, msg.ChatGUID)
 		portal.LastSeenHandle = msg.ChatGUID
 		portal.Update(nil)
 	}
 
+	// Handle message tapbacks
 	if msg.Tapback != nil {
 		portal.HandleiMessageTapback(msg)
 		return ""
-	} else if portal.bridge.DB.Message.GetLastByGUID(portal.GUID, msg.GUID) != nil {
-		portal.log.Debugln("Ignoring duplicate message", msg.GUID)
-		// Send a success confirmation since it's a duplicate message
-		overrideSuccess = true
+	}
+
+	// If the message exists in the database, handle edits or retractions
+	if dbMessage != nil && dbMessage.MXID != "" {
+
+		// Clean up the text so the length is accurate
+		msg.Text = strings.ReplaceAll(msg.Text, "\ufffc", "")
+		msg.Subject = strings.ReplaceAll(msg.Subject, "\ufffc", "")
```

Comment:
What's this for? Those placeholder characters should only be there if there's an attachment in the message, so there shouldn't be any need to remove them here

Reply:
Apple doesn't actually seem to use the retraction date for some reason, and they instead "edit" the message to remove all the content. To know whether the message has had the content removed, I needed to apply the same logic I [found elsewhere in the code](https://github.com/mautrix/imessage/blob/6118c551ac9d840e6ca2e5ab1c30efc8f60549c3/portal.go#L1847C1-L1851C3) which gave an accurate count of what the message contained. If the message is actually an attachment with this weird char, I don't want to misinterpret a delete for an edit.

Reply:
But the attachment should be deleted if the message is deleted, so there shouldn't be any weird chars?